lib/uiautomator2.js

```diff
@@ -99,116 +94,181 @@ class UiAutomator2Server {
         });
       }
     }
-    let retries = getRetries('Server install', installTimeout, SERVER_INSTALL_RETRIES);
 
-    logger.debug(`Waiting up to ${retries * 1000}ms for instrumentation '${INSTRUMENTATION_TARGET}' to be available`);
-    let output;
+    await this.verifyServicesAvailability();
+  }
+
+  async verifyServicesAvailability () {
+    log.debug(`Waiting up to ${SERVICES_LAUNCH_TIMEOUT}ms for services to be available`);
+    let isPmServiceAvailable = false;
+    let isAmServiceAvailable = false;
+    let pmOutput = '';
+    let pmError = null;
+    let amError = null;
     try {
-      await retryInterval(retries, 1000, async () => {
-        output = await this.adb.shell(['pm', 'list', 'instrumentation']);
-        if (output.indexOf('Could not access the Package Manager') !== -1) {
-          let err = new Error(`Problem running package manager: ${output}`);
-          output = null; // remove output, so it is not printed below
-          throw err;
-        } else if (output.indexOf(INSTRUMENTATION_TARGET) === -1) {
-          throw new Error('No instrumentation process found. Retrying...');
+      await waitForCondition(async () => {
+        if (!isPmServiceAvailable) {
+          pmError = null;
+          pmOutput = '';
+          try {
+            pmOutput = await this.adb.shell(['pm', 'list', 'instrumentation']);
+          } catch (e) {
+            pmError = e;
+          }
+          if (pmOutput.includes('Could not access the Package Manager')) {
+            pmError = new Error(`Problem running Package Manager: ${pmOutput}`);
+            pmOutput = ''; // remove output, so it is not printed below
+          } else if (pmOutput.includes(INSTRUMENTATION_TARGET)) {
+            pmOutput = ''; // remove output, so it is not printed below
+            log.debug(`Instrumentation target '${INSTRUMENTATION_TARGET}' is available`);
+            isPmServiceAvailable = true;
+          } else if (!pmError) {
+            pmError = new Error('The instrumentation target is not listed by Package Manager');
+          }
         }
-        logger.debug(`Instrumentation '${INSTRUMENTATION_TARGET}' available`);
+        if (!isAmServiceAvailable) {
+          // https://travis-ci.org/appium/appium-uiautomator2-driver/jobs/478514097#L4806
+          amError = null;
+          try {
+            await this.adb.shell(['am']);
```

Comment:
Is this command expected?

I failed test cases in ruby_lib_core because of this line.
https://gist.github.com/KazuCocoa/244e7cff066353be5dfc8e916d292c87

The result of `/Users/kazuaki/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am` which issued here was `255: command not found`. Then, the error caches and it becomes an error. We can see the list of command which is available `am` option.

Should we ignore the error only the result is `command not found`?

Reply:
The result of command was:

```
$ /Users/kazuaki/Library/Android/sdk/platform-tools/adb -P 5037 -s emulator-5554 shell am
Activity manager (activity) commands:
  help
      Print this help text.
  start-activity [-D] [-N] [-W] [-P <FILE>] [--start-profiler <FILE>]
          [--sampling INTERVAL] [--streaming] [-R COUNT] [-S]
          [--track-allocation] [--user <USER_ID> | current] <INTENT>
      Start an Activity.  Options are:
      -D: enable debugging
      -N: enable native debugging
      -W: wait for launch to complete
      --start-profiler <FILE>: start profiler and send results to <FILE>
      --sampling INTERVAL: use sample profiling with INTERVAL microseconds
          between samples (use with --start-profiler)
      --streaming: stream the profiling output to the specified file
          (use with --start-profiler)
      -P <FILE>: like above, but profiling stops when app goes idle
      --attach-agent <agent>: attach the given agent before binding
      -R: repeat the activity launch <COUNT> times.  Prior to each repeat,
          the top activity will be finished.
      -S: force stop the target app before starting the activity
      --track-allocation: enable tracking of object allocations
      --user <USER_ID> | current: Specify which user to run as; if not
          specified then run as the current user.
      --stack <STACK_ID>: Specify into which stack should the activity be put.
...
```

The exit status was `255: command not found` in `$?` after the command